src/dataflow/src/render/join/mod.rs

```diff
@@ -0,0 +1,275 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Rendering of `RelationExpr::Join` operators, and supporting types.
+
+pub mod delta_join;
+pub mod linear_join;
+
+use std::collections::HashMap;
+
+use expr::{MapFilterProject, ScalarExpr};
+use repr::{Datum, Row, RowArena, RowPacker};
+
+/// A manual closure implementation of filtering and logic application.
+///
+/// This manual implementation exists to express lifetime constraints clearly,
+/// as there is a relationship between the borrowed lifetime of the closed-over
+/// state and the arguments it takes when invoked. It was not clear how to do
+/// this with a Rust closure (glorious battle was waged, but ultimately lost).
+pub struct JoinClosure {
+    ready_equivalences: Vec<Vec<ScalarExpr>>,
+    before: MapFilterProject,
+}
+
+impl JoinClosure {
+    /// Applies per-row filtering and logic.
+    #[inline(always)]
+    pub fn apply<'a>(
+        &'a self,
+        datums: &mut Vec<Datum<'a>>,
+        temp_storage: &'a RowArena,
+        row_packer: &mut RowPacker,
+    ) -> Result<Option<Row>, expr::EvalError> {
+        for exprs in self.ready_equivalences.iter() {
+            // Each list of expressions should be equal to the same value.
+            let val = exprs[0].eval(&datums[..], &temp_storage)?;
+            for expr in exprs[1..].iter() {
+                if expr.eval(&datums, &temp_storage)? != val {
+                    return Ok(None);
+                }
+            }
+        }
+        self.before.evaluate(datums, &temp_storage, row_packer)
+    }
+
+    /// Construct an instance of the closure from available columns.
+    ///
+    /// This method updates the available columns, equivalences, and
+    /// the `MapFilterProject` instance. The columns are updated to
+    /// include reference to any columns added by the application of
+    /// this logic, which might result from partial application of
+    /// the `MapFilterProject` instance.
+    ///
+    /// If all columns are available for `mfp`, this method works
+    /// extra hard to ensure that the closure contains all the work,
+    /// and `mfp` is left as an identity transform (which can then
+    /// be ignored).
+    pub fn build(
```

Comment:
minor: this function looks like it doesnt need to be public (tried without it and it compiled)